nodejs/lancedb/embedding/utils.ts

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The LanceDB Authors
+
+/**
+ * Retry a function with exponential backoff.
+ * 
+ * @param func The function to be retried
+ * @param initialDelay Initial delay in seconds (default is 1)
+ * @param exponentialBase The base for exponential backoff (default is 2)
+ * @param jitter Whether to add jitter to the delay (default is true)
+ * @param maxRetries Maximum number of retries (default is 7)
+ * @returns The result of the function call
+ */
+export async function retryWithExponentialBackoff<T>(
+  func: () => Promise<T>,
+  initialDelay: number = 1,
+  exponentialBase: number = 2,
+  jitter: boolean = true,
+  maxRetries: number = 7
+): Promise<T> {
+  let numRetries = 0;
+  let delay = initialDelay;
+
+  // Loop until a successful response or max_retries is hit or an exception is raised
+  while (true) {
+    try {
+      return await func();
+    } catch (error) {
+      numRetries += 1;
+
+      if (numRetries > maxRetries) {
+        throw new Error(`Maximum number of retries (${maxRetries}) exceeded.`, { cause: error });
+      }
+
+      delay *= exponentialBase * (1 + (jitter ? Math.random() : 0));
+      console.warn(
+        `Error occurred: ${error.message}\nRetrying in ${delay.toFixed(2)} seconds (retry ${numRetries} of ${maxRetries})`
+      );
+
+      // Wait for the calculated delay
+      await new Promise((resolve) => setTimeout(resolve, delay * 1000));
```

Comment:
FYI We require Node >= 18, so you can use the Promise-native `setTimeout` API: https://nodejs.org/api/timers.html#timerspromisessettimeoutdelay-value-options